stacks/udpconn.go

```diff
@@ -0,0 +1,230 @@
+	RxBufSize uint16
+}
+
+func NewUDPConn(stack *PortStack, cfg BuffConfig) (*UDPConn, error) {
+	
+	if cfg.RxBufSize == 0 {
+			cfg.RxBufSize = defaultUDPbuffSize
+	}
+	if cfg.TxBufSize == 0 {
+		cfg.TxBufSize = defaultUDPbuffSize
+	}
+
+	buf := make([]byte, cfg.RxBufSize+cfg.TxBufSize) //I guess slicing this single buffer is a heap allocation optimisation or something - but I'm not really a fan TX and RX buffers would be a lot more readable
+	
+	sock := makeUDPConn(stack, buf[:cfg.TxBufSize], buf[cfg.TxBufSize:cfg.TxBufSize+cfg.RxBufSize])
+	sock.trace("NewUDPConn:end")
+	return &sock, nil
+
+}
+
+func makeUDPConn(stack *PortStack, tx, rx []byte) UDPConn {
+	return UDPConn{
+		stack: stack,
+		tx:    ring{buf: tx},
+		rx:    ring{buf: rx},
+	}
+}
+
+// OpenDialUDP won't really do anything - other than choose a local outbound port) ???
+func (sock *UDPConn) OpenDialUDP(localPort uint16, remoteMAC [6]byte, remote netip.AddrPort) error {
+		
+	sock.trace("UDPConn.OpenDialUDP:start")
+	return sock.openstack( localPort,remoteMAC, remote)
+
+}
+
+func (sock *UDPConn) openstack(localPortNum uint16, remoteMAC [6]byte, remote netip.AddrPort) error { //}, iss seqs.Value, remoteMAC [6]byte, remoteAddr netip.AddrPort) error {
+	
+	sock.stack.OpenUDP(localPortNum, sock)
+	err := sock.open(localPortNum,  remoteMAC, remote)
+
+	return err
+	
+}
+
+func (sock *UDPConn) open( localPortNum uint16, remoteMAC [6]byte, remoteAddr netip.AddrPort) error {
+	
+	// err := sock.scb.Open(iss, seqs.Size(len(sock.rx.buf)), state)
+	// if err != nil {
+	// 	return err
+	// }
+	// sock.scb.SetLogger(sock.stack.logger)
+	sock.remoteMAC = remoteMAC //this is our router/gateway MAC address - we never get to know the remote MAC address (that would be a security issue)
+	sock.remote = remoteAddr
+	sock.localPort = localPortNum
+	sock.rx.Reset()
+	sock.tx.Reset()
+	
+	return nil
+}
+
+
+func (u *UDPConn) abort() {
+    // There is no connection per se to abort
+	
+}
+
+func (u *UDPConn) SetReadDeadline(t time.Time) error  {
+    // not really relevant to UDP - buit the interface requires it
+	return nil
+}
+
+func (u *UDPConn) SetWriteDeadline(t time.Time) error  {
+    // not really relevant to UDP
+	return nil
+}
+
+func (sock *UDPConn) isPendingHandling() bool  {
+    
+	// much simpler than TCP - may need expanding??
+	return (sock.tx.Buffered()>0) || (sock.rx.Buffered() > 0 )
+	
+}
+
+
+// Reads from the underlying rx ring buffer, throws an EOF error if no data is available
+func (u *UDPConn) Read(b []byte) (n int, err error) {    
+	return u.rx.Read(b) //read from the rx ring buffer into b    
+}
+
+// Writes into the underlying tx ring buffer - calls to the portStacks handleEth() will send the (queued) data
+func (sock *UDPConn) Write(b []byte) (n int, err error) {    
+	err = sock.stack.FlagPendingUDP(sock.localPort)	
+	if err != nil {panic ("flagpending error")} //@@@REMOVE ?
+    return sock.tx.Write(b)
+}
+
+
+func (u *UDPConn) Close() error {    
+    return nil
```

Comment:
Close should close the connection as in PortStack should release the port for future use. As of now your Conn leaks after opening, it hogs the port forever.

Reply:
found closeUDP (well, copilot did !)